### PR TITLE
web3 connector

### LINKED ISSUE #8: Build wallet connection UI
`Edit to view comments`

## Requirement: 

```
As a marketplace user,
I want to be able to choose what type of wallet to connect with (metamask/wallet-connect),
So that I can connect with my specific wallet type
```
```
As a marketplace user,
I want to be able to connect my wallet to the dapp,
So I can use the marketplace functionality
```
```
As a marketplace user,
I want to be able to disconnect my wallet from the dapp,
So I can end my session
```
```
As a marketplace user,
I want to be able to be informed if I am on an unsupported network,
So I can change to the correct network
```
- Ability for the user to connect/disconnect their wallet to the polygon network
- Initially have two wallet options: 
   - Metamask
   - Wallet Connect
- UI informing to set to correct network

## Acceptance criteria:

- User can access the marketplace on dev server and connect their wallet via all the connection options

## Dependency:

#18 

## ETA:

- 2 days

## Links:

- https://github.com/NoahZinsmeister/web3-react
- Message Umair for the answers...

### PR DESCRIPTION
**Pull request by abbeydabiri**
Fixes #8 #53

## Description

Added web3wrapper and web3 connector.
Used @ethersproject/providers web3provider
Added wallet address to site header and implemented disconnect button
Fixed conflicts with packagejson from develop branch

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "vhc-marketplace-app",
+  "name": "vhc-markfetherplace-app",
```
Comment (anjorinjnr) [inline package.json:2]: what is this name change about?

File: package.json [hunk 2]
```diff
@@ -32,19 +33,25 @@
...
     "@types/react-redux": "^7.1.23",
+    "@web3-react/core": "^6.1.9",
+    "@web3-react/injected-connector": "^6.0.7",
     "ethers": "^5.6.1",
     "graphql": "^15.8.0",
     "graphql-request": "^4.2.0",
     "graphql-ts-client-api": "^3.0.10",
     "graphql-ts-client-codegen": "^3.0.10",
     "react": "^17.0.2",
+    "react-app-rewired": "^2.2.1",
```
Comment (anjorinjnr) [inline package.json:44]: what is this for?

File: package.json [hunk 3]
```diff
@@ -32,19 +33,25 @@
...
     "react-dom": "^17.0.2",
     "react-helmet-async": "^1.2.3",
     "react-redux": "^7.2.6",
     "react-router-dom": "^6.2.1",
     "react-scripts": "5.0.0",
+    "stream-browserify": "^3.0.0",
```
Comment (anjorinjnr) [inline package.json:50]: do we still need all these deps with ether.js?

File: src/features/partials/Popups/ConnectWallet.tsx [hunk 4]
```diff
@@ -1,16 +1,31 @@
...
 const ConnectWallet = () => {
+  const [, setWeb3Loaded] = useState(false)
+  const { activate } = useWeb3React()
+
+  async function connect() {
+    try {
+      await activate(injected)
+      setWeb3Loaded(true)
+    } catch (ex) {
+      setWeb3Loaded(true)
```
Comment (anjorinjnr) [inline src/features/partials/Popups/ConnectWallet.tsx:18]: why is this also true

Comment (anjorinjnr) [reply src/features/partials/Popups/ConnectWallet.tsx:18]: why is this  even set if it's not being used anyways

File: src/features/shared/header/Header.tsx [hunk 5]
```diff
@@ -24,105 +27,120 @@ const Header = () => {
     dispatch(openModal(Popup.CONNECT_WALLET))
   }
 
+  async function disconnect() {
+    try {
+      // eslint-disable-next-line no-console
+      console.log('deactivate')
```
Comment (anjorinjnr) [inline src/features/shared/header/Header.tsx:33]: remove the console.log

File: src/features/shared/header/Header.tsx [hunk 6]
```diff
@@ -12,7 +13,9 @@ import Logo from './Logo'
 const Header = () => {
   const activeLink = navList[0].text
   const [isClicked, setIsClicked] = useState(activeLink)
-  const [connected] = useState(false) // temporary state management for when wallet is connected and when it's not
+
+  const [, setWeb3Loaded] = useState(false)
```
Comment (anjorinjnr) [inline src/features/shared/header/Header.tsx:17]: what is this for?

File: src/web3/web3wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+}: {
+  children: JSX.Element
+}): JSX.Element {
+  const {
+    active: networkActive,
+    error: networkError,
+    activate: activateNetwork
+  } = useWeb3React()
+  const [web3loaded, setWeb3Loaded] = useState(false)
+  // const [provider] = useState('metamask')
```
Comment (anjorinjnr) [inline src/web3/web3wrapper.ts:17]: remove

File: src/web3/web3wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    injected
+      .isAuthorized()
+      .then((isAuthorized) => {
+        if (isAuthorized && !networkActive && !networkError && !web3loaded) {
+          activateNetwork(injected)
+          setWeb3Loaded(true)
+        }
+      })
+      .catch(() => {
+        setWeb3Loaded(true)
```
Comment (anjorinjnr) [inline src/web3/web3wrapper.ts:28]: why is `setWeb3Loaded` always set to true?

File: src/web3/web3wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { useEffect, useState } from 'react'
+
+import { useWeb3React } from '@web3-react/core'
+import { injected } from './connectors'
+
+export function Web3Wrapper({
```
Comment (anjorinjnr) [inline src/web3/web3wrapper.ts:6]: this doesn't actually wrap anything

I think this code can just live in the header, something like..

```tsx

const {
    active: networkActive,
    error: networkError,
    activate: activateNetwork
  } = useWeb3React()

  useEffect(() => {
  if(!account) {
    injected
      .isAuthorized()
      .then((isAuthorized) => {
        if (isAuthorized && !networkActive && !networkError) {
          activateNetwork(injected)
        }
      })
  } 
 
  }, [account, activateNetwork, networkActive, networkError])

```

**Review by anjorinjnr — CHANGES_REQUESTED** [review]
the modal doesn't go away after connecting

the disconnect is not intuitive, lets have a drop down in the avatar that says "Disconnect Wallet"  we can do this in separate PR @creativogee  or @chandni-vault-hill

File: package.json (hunk 2 above)
Comment (abbeydabiri) [reply package.json:44]: this was part of the solution being tested for web3 before switching to ethersprojects will remove

File: package.json (hunk 3 above)
Comment (abbeydabiri) [reply package.json:50]: no we do not need them, they have been taken out in my next commit

File: src/features/shared/header/Header.tsx (hunk 5 above)
Comment (abbeydabiri) [reply src/features/shared/header/Header.tsx:33]: console log removed

File: src/features/shared/header/Header.tsx (hunk 6 above)
Comment (abbeydabiri) [reply src/features/shared/header/Header.tsx:17]: this logic has been removed, new logic now checks with the wallet address to determine if connected

File: src/web3/web3wrapper.ts (hunk 8 above)
Comment (abbeydabiri) [reply src/web3/web3wrapper.ts:28]: setWeb3Loaded will not always be set to true, it is now set to false when we are unable to connect.

File: src/web3/web3wrapper.ts (hunk 9 above)
Comment (abbeydabiri) [reply src/web3/web3wrapper.ts:6]: the wrapper actually works and has been improved to watch for address changes.

The wrapper helps maintain persistent connection to the wallet address.

File: src/features/partials/Popups/ConnectWallet.tsx (hunk 4 above)
Comment (abbeydabiri) [reply src/features/partials/Popups/ConnectWallet.tsx]: this has been removed

**Pull request merged**